### PR TITLE
Fix testnet discrimination and intersection always being genesis 

### LINKED ISSUE #1453: cardano-node: Wallet re-syncs from 0% after restarting cardano-wallet-byron
# Context
 
| Information  | -                                                     |
| ---          | ---                                                   |
| Version      |`2020.3.16 (git revision: 52d2cf591a534051c39366825fcfd723f3512e30)`|
| Platform     | Linux, All?         |
| Installation | From github           |

# Steps to Reproduce

1. start cardano-wallet on cardano-node:
```
cardano-wallet-byron serve --node-socket /home/piotr/wb/cardano-node/state-node-mainnet/node.socket --mainnet --database ./mainnet
```
2. Create a wallet and make it sync (not necessarily up to 100%)
```
curl  -vX POST http://localhost:8090/v2/byron-wallets \
  -H "Content-Type: application/json; charset=utf-8" \
  -d '{
"style" : "random",
"name": "Byron1",
"mnemonic_sentence": ["rotate", "machine", "travel", "safe", "expire", "leopard", "wink", "vault", "borrow", "digital", "wisdom", "harsh"],
"passphrase": "Secure Passphrase"
}' --http1.1  |  jq
```
3. Stop and start again `cardano-wallet-byron serve`
4. List wallets
```
curl  -vX GET http://localhost:8090/v2/byron-wallets   |  jq
```

## Expected behavior

Wallet continues to sync starting from the point it left before stopping cardano-wallet-byron.

## Actual behavior
 
Wallet syncs again from 0%.
In the log there is:
```
[cardano-wallet.wallet-engine:Info:20] [2020-03-18 11:14:14.27 UTC] df01a993: Rolled back to 0.0
[cardano-wallet.wallet-engine:Info:23] [2020-03-18 11:14:14.27 UTC] 10e1a38c: Rolled back to 0.0
```

---

# Resolution 

- Order of headers submitted for intersection matters ==> #1481

---

# QA

### LINKED ISSUE #1462: cardano-node: Cannot send transaction on testnet (cardano-node + cardano-wallet-byron)
# Context
 

| Information  | -                                                     |
| ---          | ---                                                   |
| Version      | cardano-wallet: `2020.3.16 (git revision: 15d6b4e4d6a33fab19550581ddc967d2121ae537)`, cardano-node 1.8 |
| Platform     | Linux        |
| Installation | from source           |

# Steps to Reproduce

0. start cardano-node + wallet on testnet
 - `nix-build -A scripts.testnet.node -o testnet.node.local && testnet.node.local`
 -  `cardano-wallet-byron serve --node-socket ~/wb/cardano-node/state-node-testnet/node.socket --testnet /nix/store/7106qjwral616j6p5049m7hhi5ih0r33-testnet-genesis.json --database ./testnet`

1. Restore wallet with funds.
```
curl  -vX POST http://localhost:8090/v2/byron-wallets \
  -H "Content-Type: application/json; charset=utf-8" \
  -d '{
"style": "random",
"name": "Byron",
"mnemonic_sentence": ["recycle","august","horse","awesome","shuffle","agent","almost","cheap","fury","speak","duty","enemy"],
"passphrase": "Secure Passphrase"
}' --http1.1  |  jq
```
2. Run TX:
```
curl  -vX POST http://localhost:8090/v2/byron-wallets/d3358ac0fc699784af021710e7aba2a3e31fbbd1/transactions \
  -H "Accept: application/json; charset=utf-8" \
  -H "Content-Type: application/json; charset=utf-8" \
  -d '{
	"payments": [
		{
			"address": "37btjrVyb4KFcfEEYyBU8j4VazP5he8iktF5uyD95avRpUHU6aWuLhuhGMmyFiJjtinJQPSjgGmh5w8qmuxDpHjK2JmCnPALuA4dyUkMpHF5Rf4PBd",
			"amount": {
				"quantity": 1,
				"unit": "lovelace"
			}
		}
		],
	"passphrase": "Secure Passphrase"
}' --http1.1  |  jq
```

## Expected behavior

Transaction is successful
## Actual behavior
 
Response:
```
{
  "code": "network_unreachable",
  "message": "The node backend is unreachable at the moment. Trying again in a bit might work."
}


```

cardano-wallet log:
```
[cardano-wallet.api-server:Warning:21567] [2020-03-19 17:06:52.77 UTC] [RequestId 96] 503 Service Unavailable in 60.54625479s

```

cardano-node log:
```
[piotr-la:cardano.node.metrics:Critical:115] [2020-03-19 17:35:22.87 UTC] txsInMempool = 0
[piotr-la:cardano.node.metrics:Critical:115] [2020-03-19 17:35:22.87 UTC] txsProcessed = 1
[piotr-la:cardano.node.metrics:Critical:115] [2020-03-19 17:35:22.87 UTC] mempoolBytes = 0
[piotr-la:cardano.node.Mempool:Info:115] [2020-03-19 17:35:22.87 UTC] {"kind":"TraceMempoolRejectedTxs","mempoolSize":"MempoolSize {msNumTxs = 0, msNumBytes = 0}","txRejected":"tx: Tx 4fa11b11460c93b1 with inputs [TxInUtxo 031e05b2 #1], outputs: [TxOut 40 lovelace -> 37btjrVyb4KFcfEEYyBU8j4VazP5he8iktF5uyD95avRpUHU6aWuLhuhGMmyFiJjtinJQPSjgGmh5w8qmuxDpHjK2JmCnPALuA4dyUkMpHF5Rf4PBd, TxOut 140199068 lovelace -> 2w1sdSJu3GVfNEJF5765gtMMP5koCgkxctKFdjhEh6S54B4zpDn1Rm7VL4wpYZEa5J5AfGyn93uxMTh1oP6AApYZ8kJNbBwbxJT]\nwitnesses: [\n    VKWitness: key = pub:cdf70c1e, key hash = 815fc0f5, sig = <signature>\n]"}
[piotr-la:cardano.node.Mempool.text:Info:115] [2020-03-19 17:35:22.87 UTC] {"kind":"TraceMempoolRejectedTxs","mempoolSize":"MempoolSize {msNumTxs = 0, msNumBytes = 0}","txRejected":"tx: Tx 4fa11b11460c93b1 with inputs [TxInUtxo 031e05b2 #1], outputs: [TxOut 40 lovelace -> 37btjrVyb4KFcfEEYyBU8j4VazP5he8iktF5uyD95avRpUHU6aWuLhuhGMmyFiJjtinJQPSjgGmh5w8qmuxDpHjK2JmCnPALuA4dyUkMpHF5Rf4PBd, TxOut 140199068 lovelace -> 2w1sdSJu3GVfNEJF5765gtMMP5koCgkxctKFdjhEh6S54B4zpDn1Rm7VL4wpYZEa5J5AfGyn93uxMTh1oP6AApYZ8kJNbBwbxJT]\nwitnesses: [\n    VKWitness: key = pub:cdf70c1e, key hash = 815fc0f5, sig = <signature>\n]"}
[piotr-la:cardano.node.LocalErrorPolicy:Error:55] [2020-03-19 17:35:23.87 UTC] {"event":"ErrorPolicyUnhandledApplicationException (MuxError {errorType = MuxBearerClosed, errorMsg = \"<socket: 30> closed when reading data, waiting on next header True\", errorStack = []})","kind":"ErrorPolicyTrace","address":"LocalAddress {getFilePath = \"\"}"}

```

---

# Resolution 

- [x] Resurrect (and upgrade) golden tests for signed txs from http-bridge / cardano-sl ==> #1463
- [x] Log messages from the local tx submission protocol ==> #1461 
- [x] Analyze log messages and figure out what's wrong.
- [x] Remove hard-coded network magic to mainnet in the API Server ==> #1481 
---

# QA

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

This both fixes #1462 and fixes #1453 

# Overview

- cbb683065f3397fd81b252d0c89b2605b877da50
  :round_pushpin: **fix GenTx deserialization for Byron**
  This was actually plain wrong. Our 'SealedTx' serialized type actually corresponds
to a serialized 'TxAux' from the cardano-ledger code base. So, we can deserialize
the 'SealedTx' to this and then, work our way to 'GenTx ByronBlock' easily.

- 78f6dfa7888ef2dad7f7e4c2b03af1eba833395b
  :round_pushpin: **fix finding intersection with node**
  The node actually looks at the points from the 'FindIntersect' message
as if they were send with the most-preferred first. So, sending
the genesis point first will always result in finding an intersection
with the genesis point, regardless of what's sent after.

- 8136e029563a12e5445aacb13d8064d86ece2519
  :round_pushpin: **add extra unit tests comparing toGenTx with existing golden**
  
- 4fb9f4712ba695bd2eb91bb4ad3de7dfe45649a3
  :round_pushpin: **fix transaction submission error deserialization**
  The error returned by the protocol used to be a string in the early days, but it is now a
proper error data-type. We can't actually deserialize it as a string.

- 43f4c6b522a9ec7701a5d9753e840e0d71154637
  :round_pushpin: **remove remaining hard-coded 'Mainnet constraints all over the API layer**
  So this is the main reason why things were not working properly on testnet ... :man_facepalming: we were actually
generating mainnet change addresses, because the network discrimination was hard-coded to mainnet in both byron
api layers! This works better now, in particular the scenario described in #1462

# Comments

### CODE REVIEW

**Comment by KtorZ** [comment]
bors try

File: lib/byron/src/Cardano/Wallet/Byron/Network.hs [hunk 1]
```diff
@@ -237,17 +253,22 @@ withNetworkLayer tr bp addrInfo versionData action = do
         chainSyncQ <- atomically newTQueue
         let client = const $ mkNetworkClient tr bp chainSyncQ localTxSubmissionQ
         link =<< async (connectClient tr client versionData addrInfo)
-        let points = genesisPoint : (toPoint getEpochLength <$> headers)
+        let points = reverse $ genesisPoint : (toPoint getEpochLength <$> headers)
```
Comment (paweljakubas) [inline lib/byron/src/Cardano/Wallet/Byron/Network.hs:256]: :+1:

**Review by paweljakubas — APPROVED** [review]
looks fine!

**Pull request merged**